src/askvm/resources/core/forIn.ts

```diff
@@ -1,11 +1,52 @@
 import { any, resource, run, runUntyped } from '../../lib';
+import { isAskCode } from '../../../askcode';
 
 export const forIn = resource({
   type: any,
   async compute(options, code, args) {
     const { params } = code;
-    const [condition, args2, block] = params!;
-    const value = await runUntyped(options, condition);
-    return run(options, value ? block : null, []);
+    if (!params || params?.length !== 3) {
+      throw new Error('forIn expects exactly 3 params');
+    }
+    const [key, of, block] = params;
+    if (key === null || of === null || block === null) {
+      throw new Error('forIn expects non-null key, in and block');
+    }
+    if (!isAskCode(key)) {
+      throw new Error('forIn expects key to be a statement');
+    }
+    if (!['let', 'const'].includes(key.name)) {
+      throw new Error('forIn expects key to be a "let" or "const" statement');
+    }
+    if (!Array.isArray(key.params)) {
+      throw new Error('forIn expects key to have params!');
+    }
+
+    const valueOf = await runUntyped(options, of);
+
+    if (
+      valueOf === null ||
+      typeof valueOf == 'boolean' ||
+      typeof valueOf == 'number'
+    ) {
+      throw new Error(
+        'forIn expects "in" to be an array, an object or a string'
+      );
+    }
+
+    const values: any[] =
+      typeof valueOf === 'object'
+        ? Object.keys(valueOf)
+        : Array.from({ length: valueOf.length }, (value, key) => key);
+
+    for (const value of values) {
+      key.params[1] = value;
```

Comment:
it's a little hacky to modify the param like here, but it's okay for the first implementation fo this resource.

Reply:
Thanks. By the way, I think I did similar for for of.